README.md
[![CMake](https://github.com/kuvaus/LlamaGPTJ-chat/actions/workflows/cmake.yml/badge.svg)](https://github.com/kuvaus/LlamaGPTJ-chat/actions/workflows/cmake.yml)
# LlamaGPTJ-chat
Simple command line chat program for [GPT-J](https://en.wikipedia.org/wiki/GPT-J), [LLaMA](https://en.wikipedia.org/wiki/LLaMA) and [MPT](https://www.mosaicml.com/blog/mpt-7b) models written in C++. Based on [llama.cpp](https://github.com/ggerganov/llama.cpp) and uses [gpt4all-backend](https://github.com/nomic-ai/gpt4all) for full compatibility.

<img alt="LlamaGPTJ-chat demo" src="https://user-images.githubusercontent.com/22169537/234323778-64365dc9-8bd9-4a48-b7de-ec0280a5fb4e.gif" width="600" />

> **Warning**
> Very early progress, might have bugs

# Table of contents
<!-- TOC -->
* [Installation](#installation)
* [Usage](#usage)
* [GPT-J, LLaMA, and MPT models](#gpt-j-llama-and-mpt-models)
* [Detailed command list](#detailed-command-list)
* [Useful features](#useful-features)
* [License](#license)
<!-- TOC -->

## Installation
Since the program is made using c++ it should build and run on most Linux, MacOS and Windows systems. The [Releases](https://github.com/kuvaus/LlamaGPTJ-chat/releases) link has ready-made binaries. AVX2 is faster and works on most newer computers. If you run the program, it will check and print if your computer has AVX2 support.

### Download
```sh
git clone --recurse-submodules https://github.com/kuvaus/LlamaGPTJ-chat
cd LlamaGPTJ-chat
```
You need to also download a model file, see [supported models](#gpt-j-llama-and-mpt-models) for details and links.

### Build
Since the program is made using c++ it should build and run on most Linux, MacOS and Windows systems. 
On most systems, you only need this to build:
```sh
mkdir build
cd build
cmake ..
cmake --build . --parallel
```
> **Note**
> 
> If you have an old processor, you can turn AVX2 instructions OFF in the build step with `-DAVX2=OFF` flag.
> 
> If you have a new processor, you can turn AVX512 instructions ON in the build step with `-DAVX512=ON` flag.
> 
> On old macOS, set `-DBUILD_UNIVERSAL=OFF` to make the build x86 only instead of the universal Intel/ARM64 binary.
> On really old macOS, set `-DOLD_MACOS=ON`. This disables `/save` and `/load` but compiles on old Xcode.
> 
> On Windows you can now use Visual Studio (MSVC) or MinGW. If you want MinGW build instead, set `-G "MinGW Makefiles"`.
>
> On ARM64 Linux there are no ready-made binaries, but you can now build it from source.

## Usage

After compiling, the binary is located at:

```sh
build/bin/chat
```
But you're free to move it anywhere. Simple command for 4 threads to get started:
```sh
./chat -m "/path/to/modelfile/ggml-vicuna-13b-1.1-q4_2.bin" -t 4
```
or
```sh
./chat -m "/path/to/modelfile/ggml-gpt4all-j-v1.3-groovy.bin" -t 4
```

Happy chatting!


## GPT-J, LLaMA, and MPT models
Current backend supports the GPT-J, LLaMA and MPT models.

### GPT-J model
You need to download a GPT-J model first. Here are direct links to models:

>- The default version is **v1.0**: [ggml-gpt4all-j.bin](https://gpt4all.io/models/ggml-gpt4all-j.bin)
>- At the time of writing the newest is **1.3-groovy**: [ggml-gpt4all-j-v1.3-groovy.bin](https://gpt4all.io/models/ggml-gpt4all-j-v1.3-groovy.bin)

They're around 3.8 Gb each. The chat program stores the model in RAM on runtime so you need enough memory to run. You can get more details on GPT-J models from [gpt4all.io](https://gpt4all.io/) or [nomic-ai/gpt4all](https://github.com/nomic-ai/gpt4all) github.

### LLaMA model
Alternatively you need to download a LLaMA model first. The original weights are for research purposes and you can apply for access [here](https://ai.facebook.com/blog/large-language-model-llama-meta-ai/). Below are direct links to derived models:

>- Vicuna 7b **v1.1**: [ggml-vicuna-7b-1.1-q4_2.bin](https://gpt4all.io/models/ggml-vicuna-7b-1.1-q4_2.bin)
>- Vicuna 13b **v1.1**: [ggml-vicuna-13b-1.1-q4_2.bin](https://gpt4all.io/models/ggml-vicuna-13b-1.1-q4_2.bin)
>- GPT-4-All **l13b-snoozy**: [ggml-gpt4all-l13b-snoozy.bin](https://gpt4all.io/models/ggml-gpt4all-l13b-snoozy.bin)

The LLaMA models are quite large: the 7B parameter versions are around 4.2 Gb and 13B parameter 8.2 Gb each. The chat program stores the model in RAM on runtime so you need enough memory to run. You can get more details on LLaMA models from the [whitepaper](https://arxiv.org/abs/2302.13971) or META AI [website](https://ai.facebook.com/blog/large-language-model-llama-meta-ai/).

### MPT model
You can also download and use an MPT model instead. Here are direct links to MPT-7B models:
>- MPT-7B base model pre-trained by Mosaic ML: [ggml-mpt-7b-base.bin](https://gpt4all.io/models/ggml-mpt-7b-base.bin)
>- MPT-7B instruct model trained by Mosaic ML: [ggml-mpt-7b-instruct.bin](https://gpt4all.io/models/ggml-mpt-7b-instruct.bin)
>- Non-commercial MPT-7B chat model  trained by Mosaic ML: [ggml-mpt-7b-chat.bin](https://gpt4all.io/models/ggml-mpt-7b-chat.bin)

They're around 4.9 Gb each. The chat program stores the model in RAM on runtime so you need enough memory to run. You can get more details on MPT models from MosaicML [website](https://www.mosaicml.com/blog/mpt-7b) or [mosaicml/llm-foundry](https://github.com/mosaicml/llm-foundry) github.

## Detailed command list
You can view the help and full parameter list with:
`
./chat -h
`

```sh
usage: ./bin/chat [options]

A simple chat program for GPT-J, LLaMA, and MPT models.
You can set specific initial prompt with the -p flag.
Runs default in interactive and continuous mode.
Type '/reset' to reset the chat context.
Type '/save','/load' to save network state into a binary file.
Type '/save NAME','/load NAME' to rename saves. Default: --save_name NAME.
Type '/help' to show this help dialog.
Type 'quit', 'exit' or, 'Ctrl+C' to quit.

options:
  -h, --help            show this help message and exit
  -v, --version         show version and license information
  --run-once            disable continuous mode
  --no-interactive      disable interactive mode altogether (uses given prompt only)
  --no-animation        disable chat animation
  --no-saves            disable '/save','/load' functionality
  -s SEED, --seed SEED  RNG seed for --random-prompt (default: -1)
  -t N, --threads    N  number of threads to use during computation (default: 4)
  -p PROMPT, --prompt PROMPT
                        prompt to start generation with (default: empty)
  --random-prompt       start with a randomized prompt.
  -n N, --n_predict  N  number of tokens to predict (default: 200)
  --top_k            N  top-k sampling (default: 40)
  --top_p            N  top-p sampling (default: 0.9)
  --temp             N  temperature (default: 0.9)
  --n_ctx            N  number of tokens in context window (default: 0)
  -b N, --batch_size N  batch size for prompt processing (default: 20)
  --repeat_penalty   N  repeat_penalty (default: 1.1)
  --repeat_last_n    N  last n tokens to penalize  (default: 64)
  --context_erase    N  percent of context to erase  (default: 0.8)
  --b_token             optional beginning wrap token for response (default: empty)
  --e_token             optional end wrap token for response (default: empty)
  -j,   --load_json FNAME
                        load options instead from json at FNAME (default: empty/no)
  --load_template   FNAME
                        load prompt template from a txt file at FNAME (default: empty/no)
  --save_log        FNAME
                        save chat log to a file at FNAME (default: empty/no)
  --load_log        FNAME
                        load chat log from a file at FNAME (default: empty/no)
  --save_dir        DIR
                        directory for saves (default: ./saves)
  --save_name       NAME
                        save/load model state binary at save_dir/NAME.bin (current: model_state)
                        context is saved to save_dir/NAME.ctx (current: model_state)
  -m FNAME, --model FNAME
                        model path (current: ./models/ggml-vicuna-13b-1.1-q4_2.bin)
```
## Useful features
Here are some handy features and details on how to achieve them using command line options.

### Save/load chat log and read output from other apps
By default, the program prints the chat to standard (stdout) output, so if you're including the program into your app, it only needs to read stdout. You can also save the whole chat log to a text file with `--save_log` option. There's an elementary way to remember your past conversation by simply loading the saved chat log with `--load_log` option when you start a new session.

### Run the program once without user interaction
If you only need the program to run once without any user interactions, one way is to set prompt with `-p "prompt"` and using `--no-interactive` and `--no-animation` flags. The program will read the prompt, print the answer, and close.

### Add AI personalities and characters
If you want a personality for your AI, you can change `prompt_template_sample.txt` and use `--load_template` to load the modified file. The only constant is that your input during chat will be put on the `%1` line. Instructions, prompt, response, and everything else can be replaced any way you want. Having different `personality_template.txt` files is an easy way to add different AI characters. With _some_ models, giving both AI and user names instead of `Prompt:` and `Response:`, can make the conversation flow more naturally as the AI tries to mimic a conversation between two people.

### Ability to reset chat context
You can reset the chat at any time during chatting by typing `/reset` in the input field. This will clear the AI's memory of past conversations, logits, and tokens. You can then start the chat from a blank slate without having to reload the whole model again.

### Load all parameters using JSON
You can also fetch parameters from a json file with `--load_json "/path/to/file.json"` flag. Different models might perform better or worse with different input parameters so using json files is a handy way to store and load all the settings at once. The JSON file loader is designed to be simple in order to prevent any external dependencies, and as a result, the JSON file must follow a specific format. Here is a simple example:

```javascript
{"top_p": 1.0, "top_k": 50400, "temp": 0.9, "n_batch": 9}
```
This is useful when you want to store different temperature and sampling settings.

And a more detailed one:
```javascript
{
"top_p": 1.0,
"top_k": 50400,
"temp": 0.9,
"n_batch": 20,
"threads": 12,
"prompt": "Once upon a time",
"load_template": "/path/to/prompt_template_sample.txt",
"model": "/path/to/ggml-gpt4all-j-v1.3-groovy.bin",
"no-interactive": "true"
}
```
This one loads the prompt from the json, uses a specific template, and runs the program once in no-interactive mode so user does not have to press any input.

## License

This project is licensed under the MIT [License](https://github.com/kuvaus/LlamaGPTJ-chat/blob/main/LICENSE)


CHANGELOG.md

## Changelog

#### [Upcoming](https://github.com/kuvaus/LlamaGPTJ-chat/compare/v0.3.0...HEAD)


#### [v0.3.0](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.3.0)

> 26 June 2023

- Add this [changelog](https://github.com/kuvaus/LlamaGPTJ-chat/blob/main/CHANGELOG.md) :)
- Add sha256 hashes on release so you can verify the binaries
- All binaries are automatically generated with Github actions
- Add signal handling for SIGHUP (macOS, Linux) and CTRL_CLOSE_EVENT (Windows) to fix issue [`#16`](https://github.com/kuvaus/LlamaGPTJ-chat/issues/16)
- This allows you to run chat as a subprocess. The chat subprocess now quits properly if parent app is closed.
- Version information
- Fix segfault on`/help`

#### [v0.2.9](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.9)

> 22 June 2023

- [Pull request](https://github.com/kuvaus/LlamaGPTJ-chat/pull/18) from [@154pinkchairs](https://github.com/154pinkchairs/) merged. Thanks. :)
- The pull request [`#18`](https://github.com/kuvaus/LlamaGPTJ-chat/pull/18) has the two fixes below:
- Properly handle file paths including tildes [`18e9f36`](https://github.com/kuvaus/LlamaGPTJ-chat/commit/18e9f36)
- Handle buffer allocation errors [`6800dfb`](https://github.com/kuvaus/LlamaGPTJ-chat/commit/6800dfb)
- Better debug mode compilation. May fix issue [`#9`](https://github.com/kuvaus/LlamaGPTJ-chat/issues/9)

#### [v0.2.8](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.8)
> 16 June 2023

- Adds `--save_dir` option so you can change save directory location
- Default location is `./saves` on the same directory as the chat binary
- See issue [`#13`](https://github.com/kuvaus/LlamaGPTJ-chat/issues/13) for more details

#### [v0.2.7](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.7)
> 15 June 2023

- Fixes for old macOS.
- Use `-DOLD_MACOS=ON` option when compiling with CMake.
- Tested to compile on High Sierra and Xcode 10

#### [v0.2.6](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.6)
> 14 June 2023

- You can name saves with `./save NAME` and `./load NAME`
- You can toggle saving and loading off with `--no-saves` flag

#### [v0.2.5](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.5)
> 13 June 2023

- Save/load state with `./save` and `./load`
- Reset context with `./reset`, help with `./help`
- Makes a `./saves` folder
- Note that a single save can take up to 2Gb
- You can wrap the AI response with tokens using `--b_token` and `--e_token`
- See issue [`#12`](https://github.com/kuvaus/LlamaGPTJ-chat/issues/12) for more details

#### [v0.2.4](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.4)
> 5 June 2023

- Fix when using json to specify names for logfiles. Fixes issue [`#11`](https://github.com/kuvaus/LlamaGPTJ-chat/issues/11)

#### [v0.2.3](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.3)
> 4 June 2023

- Fix said ability to reset context... :)

#### [v0.2.2](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.2)
> 3 June 2023

- Ability to reset context

#### [v0.2.1](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.1)
> 30 May 2023

- Save and load chat logs
- Use `--save_log` and `--load_log`
- AVX512 option for compilation `-DAVX512=ON`

#### [v0.2.0](https://github.com/kuvaus/LlamaGPTJ-chat/releases/tag/v0.2.0)
> 17 May 2023

- Update gpt4all backend to v0.1.1 [`61a963a`](https://github.com/kuvaus/LlamaGPTJ-chat/commit/61a963a3d220ef157a8504ddde708f33dc2946eb)
- Full Windows Visual Studio compatibility. Finally fixes issue [`#1`](https://github.com/kuvaus/LlamaGPTJ-chat/issues/1)
- Builds from source on aarch64 Linux. Fixes issue [`#3`](https://github.com/kuvaus/LlamaGPTJ-chat/issues/3)
- Full MPT support. Fixes issue [`#4`](https://github.com/kuvaus/LlamaGPTJ-chat/issues/4)

#### v0.1.9
> 16 May 2023

- Code cleaning and reordering
- `llmodel_create_model()` function

#### v0.1.8
> 13 May 2023

- Add support for MPT models
- Uses [gpt4all-backend](https://github.com/nomic-ai/gpt4all)

#### v0.1.7
> 12 May 2023

- First [pull request](https://github.com/kuvaus/LlamaGPTJ-chat/pull/2)  from [@itz-coffee](https://github.com/itz-coffee/) merged. Thanks. :)
- The pull request [`#2`](https://github.com/kuvaus/LlamaGPTJ-chat/pull/2) adds the feature below:
- Add --no-animation flag [`fdc2ac3`](https///github.com/kuvaus/LlamaGPTJ-chat/commit/fdc2ac3)
- Support for old macOS

#### v0.1.6
> 4 May 2023

- Parse parameters from json files
- Use `-j FNAME` or`--load_json  FNAME`

#### v0.1.5
> 3 May 2023

- MinGW compilation on Windows

#### v0.1.4
> 1 May 2023

- v0.1.4 had no tags
- It was part of `cmake-release.yml` rewrite to enable MinGW [`e7e1ebf`](https://github.com/kuvaus/LlamaGPTJ-chat/commit/e7e1ebf97d696d069bbc0ae7f0ed078739fb6642)

#### v0.1.3
> 1 May 2023

- Add loading of prompt template files
- Use `--load_template` for loading
- See `prompt_template_sample.txt` for a sample


#### v0.1.2
> 30 April 2023

- Automatic memory handling for the model

#### v0.1.1
> 29 April 2023

- Windows compilation fixes

#### v0.1.0
> 29 April 2023


- Before this, progress was in  [GPTJ-chat](https://github.com/kuvaus/GPTJ-chat/) and [Llama-chat](https://github.com/kuvaus/Llama-chat/)
- First version


src/utils.h
#pragma once

#ifndef UTILS_H
#define UTILS_H

#include "header.h" 

//Need this for Windows colors
#ifdef _WIN32
    #include <windows.h> 
#endif

bool containsSubstring(const std::string &str, const std::string &substr) {
    return str.find(substr) != std::string::npos;
}

void check_avx_support_at_startup() {
#if defined(__x86_64__) || defined(__i386__)
    const bool avx(__builtin_cpu_supports("avx"));
    const bool avx2(__builtin_cpu_supports("avx2"));
    const bool avx512(__builtin_cpu_supports("avx512f"));
    const bool fma(__builtin_cpu_supports("fma"));
    if (avx512 && avx && avx2 && fma) {std::cout << "Your computer supports AVX512" << std::endl;}
    else if (avx && avx2 && fma)      {std::cout << "Your computer supports AVX2" << std::endl;}
    else if (avx)                     {std::cout << "Your computer only supports AVX1" << std::endl;}
    else                    {std::cout << "Your computer does not support AVX1 or AVX2\nThe program will likely not run." << std::endl;} 
    #ifdef OLD_MACOS
    std::cout << "Compiled with OLD_MACOS flag. /save and /load features turned off." << std::endl;
    #endif
#endif
}


//////////////////////////////////////////////////////////////////////////
////////////                 SIGNAL HANDLING                  ////////////
//////////////////////////////////////////////////////////////////////////



volatile sig_atomic_t sighup_received = 0;

void handle_sighup(int signal) {
    #ifndef _WIN32
    if (signal == SIGHUP) {
        sighup_received = 1;
    }
    #endif
}

#ifdef _WIN32
BOOL WINAPI console_ctrl_handler(DWORD ctrl_type) {
    switch (ctrl_type) {
        case CTRL_C_EVENT:
        case CTRL_CLOSE_EVENT:
            sighup_received = 1;
            return TRUE;
        default:
            return FALSE;
    }
}
#endif

//////////////////////////////////////////////////////////////////////////
////////////                /SIGNAL HANDLING                  ////////////
//////////////////////////////////////////////////////////////////////////

//////////////////////////////////////////////////////////////////////////
////////////            READ PROMPT TEMPLATE FILE             ////////////
//////////////////////////////////////////////////////////////////////////

//This is a bit messy function but it should parse the template file into prefix, header, and footer.
//Chat will then prompt the model with (prefix + header + input/prompt +  footer)
std::tuple<std::string, std::string, std::string> read_prompt_template_file(const std::string& file_path) {
    std::string prefix, header, footer;
    std::ifstream file(file_path);

    std::vector<std::string> lines;
    std::string line;

    //store all lines of header template into a vector
    if (file.is_open()) {
        while (std::getline(file, line)) {
            lines.push_back(line);
        }
        file.close();
    } else {
        std::cerr << "Unable to open the prompt template file." << std::endl;
        std::cerr << "Reverting to default prompt template." << std::endl;
        return std::make_tuple("### Instruction:\n The prompt below is a question to answer, a task to complete, or a conversation to respond to; decide which and write an appropriate response.", "\n### Prompt: ", "\n### Response: "); 
    }

    //find line containing %1 and store its index.
    int input_index;
    for (size_t i = 0; i < lines.size(); ++i) {
        if (lines[i].find("%1") != std::string::npos) {
            input_index = i;
        }
    }
    //Special case of having only %1 in template file.
    if (input_index == 0) {
        header = "";
        prefix = "";
        footer = "";
    //If there is only 1 line above %1, that will be ### header.
    } else if (input_index == 1) {
        header = lines[0];
        prefix = " ";
    } else {
        
        //Put lines above the header-line into prefix.
        prefix = lines[0];
        for (size_t i = 1; i < input_index-1; ++i) {
            prefix = prefix + "\n" + lines[i];
        }
        prefix = prefix  + " ";

        //store header-line (line above input-line)
        header = "\n" + lines[input_index-1] + " ";

        //Put lines below the input-line into footer.
        footer = "\n";
        for (size_t i = input_index+1; i < lines.size(); ++i) {
             footer = footer + lines[i]+" ";
        }
    }

    return std::make_tuple(prefix, header, footer);
}


//////////////////////////////////////////////////////////////////////////
////////////           /READ PROMPT TEMPLATE FILE             ////////////
//////////////////////////////////////////////////////////////////////////

void save_chat_log(std::string save_log, std::string prompt, std::string answer) {
  std::ofstream logfile(save_log, std::ios::app);
  if (logfile.is_open()) {
    logfile << prompt;
    logfile << answer+"\n";
    logfile.close();
    }
}


std::string read_chat_log(std::string load_log) {

    std::ifstream ifs(load_log);
    std::string content((std::istreambuf_iterator<char>(ifs)),
                         std::istreambuf_iterator<char>());
    return content;
}

std::string pathname_directory(const std::string &pathname)
{
    std::size_t len = pathname.find_last_of("/\\");
    return len == std::string::npos ? "": pathname.substr(0, len);
}


void set_console_color(ConsoleState &con_st, ConsoleColor color) {
    if (con_st.use_color && con_st.color != color) {
        //Windows handles colors differently.
        #ifdef _WIN32
          WORD windows_colors[] = {
            7, 14, 10, 15
        };          
            HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
            SetConsoleTextAttribute(hConsole, windows_colors[color]);
        #else
        //ANSI colors, works for unix.
        const char* ansi_colors[] = {
            //DEFAULT, PROMPT, USER_INPUT,   BOLD
            //default, yellow, bright_green, bold
            "\x1b[0m", "\x1b[33m", "\x1b[1m\x1b[32m", "\x1b[1m"
        };
        printf("%s", ansi_colors[color]);
        #endif
        con_st.color = color;
    }
}

std::string random_prompt(int32_t seed) {
    const std::vector<std::string> prompts = {
        "So", "Once upon a time", "When", "The", "After", "If", "import", "He", "She", "They"
    };

    std::mt19937 rng(seed);
    return prompts[rng() % prompts.size()];
}

void print_version() {
	//Version/about page
	//Contains License information for distributions in binary form
std::string mit_license = R"(MIT License

Big thanks to contributors, testers, and commenters on Github.
And to you, dear user!

Happy chatting! :)
)";
std::cout << "\n\n" << APPNAME << " version " << VERSION << "\n\n" << "Made by kuvaus" << "\n\n" << mit_license << std::endl; 
//std::cout << mit_license << std::endl; 
}

void print_usage(int argc, char** argv, const chatParams& params) {
    // Print usage information
    fprintf(stderr, "usage: %s [options]\n", argv[0]);
    fprintf(stderr, "\n");
    fprintf(stderr, "A simple chat program for GPT-J, LLaMA, and MPT models.\n");
    fprintf(stderr, "You can set specific initial prompt with the -p flag.\n");
    fprintf(stderr, "Runs default in interactive and continuous mode.\n");
    fprintf(stderr, "Type '/reset' to reset the chat context.\n");
    fprintf(stderr, "Type '/save','/load' to save network state into a binary file.\n");
    fprintf(stderr, "Type '/save NAME','/load NAME' to rename saves. Default: --save_name NAME.\n");
    fprintf(stderr, "Type '/help' to show this help dialog.\n");
    fprintf(stderr, "Type 'quit', 'exit' or, 'Ctrl+C' to quit.\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "options:\n");
    fprintf(stderr, "  -h, --help            show this help message and exit\n");
    fprintf(stderr, "  -v, --version         show version and license information\n");
    fprintf(stderr, "  --run-once            disable continuous mode\n");
    fprintf(stderr, "  --no-interactive      disable interactive mode altogether (uses given prompt only)\n");
    fprintf(stderr, "  --no-animation        disable chat animation\n");
    fprintf(stderr, "  --no-saves            disable '/save','/load' functionality\n");
    fprintf(stderr, "  -s SEED, --seed SEED  RNG seed for --random-prompt (default: -1)\n");
    fprintf(stderr, "  -t N, --threads    N  number of threads to use during computation (default: %d)\n", params.n_threads);
    fprintf(stderr, "  -p PROMPT, --prompt PROMPT\n");
    fprintf(stderr, "                        prompt to start generation with (default: empty)\n");
    fprintf(stderr, "  --random-prompt       start with a randomized prompt.\n");
    fprintf(stderr, "  -n N, --n_predict  N  number of tokens to predict (default: %d)\n", params.n_predict);
    fprintf(stderr, "  --top_k            N  top-k sampling (default: %d)\n", params.top_k);
    fprintf(stderr, "  --top_p            N  top-p sampling (default: %.1f)\n", params.top_p);
    fprintf(stderr, "  --temp             N  temperature (default: %.1f)\n", params.temp);
    fprintf(stderr, "  --n_ctx            N  number of tokens in context window (default: %d)\n", params.n_ctx);
    fprintf(stderr, "  -b N, --batch_size N  batch size for prompt processing (default: %d)\n", params.n_batch);
    fprintf(stderr, "  --repeat_penalty   N  repeat_penalty (default: %.1f)\n", params.repeat_penalty);
    fprintf(stderr, "  --repeat_last_n    N  last n tokens to penalize  (default: %d)\n", params.repeat_last_n);
    fprintf(stderr, "  --context_erase    N  percent of context to erase  (default: %.1f)\n", params.context_erase);
    fprintf(stderr, "  --b_token             optional beginning wrap token for response (default: empty)\n");
    fprintf(stderr, "  --e_token             optional end wrap token for response (default: empty)\n");
    fprintf(stderr, "  -j,   --load_json FNAME\n");
    fprintf(stderr, "                        load options instead from json at FNAME (default: empty/no)\n");
    fprintf(stderr, "  --load_template   FNAME\n");
    fprintf(stderr, "                        load prompt template from a txt file at FNAME (default: empty/no)\n");
    fprintf(stderr, "  --save_log        FNAME\n");
    fprintf(stderr, "                        save chat log to a file at FNAME (default: empty/no)\n");
    fprintf(stderr, "  --load_log        FNAME\n");
    fprintf(stderr, "                        load chat log from a file at FNAME (default: empty/no)\n");
    fprintf(stderr, "  --save_dir        DIR\n");
    fprintf(stderr, "                        directory for saves (default: %s/saves)\n", pathname_directory(argv[0]).c_str());
    fprintf(stderr, "  --save_name       NAME\n");
    fprintf(stderr, "                        save/load model state binary at save_dir/NAME.bin (current: %s)\n", params.save_name.c_str());
    fprintf(stderr, "                        context is saved to save_dir/NAME.ctx (current: %s)\n", params.save_name.c_str());
    fprintf(stderr, "  -m FNAME, --model FNAME\n");
    fprintf(stderr, "                        model path (current: %s)\n", params.model.c_str());
    fprintf(stderr, "\n");
}

bool parse_params(int argc, char** argv, chatParams& params) { 

    // Parse command-line arguments
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];

        if (arg == "-j" || arg == "--load_json") {
            params.load_json = argv[++i];
            if (!params.load_json.empty()) {
                std::cout << APPNAME << ": parsing options from json: " << params.load_json << std::endl;
                get_params_from_json(params);
            } else {
                std::cout << APPNAME << ": trying to parse options from json but got empty filename." << std::endl;
            }
        } else if (arg == "--run-once") {
            params.run_once = true;
        } else if (arg == "--no-interactive") {
            params.no_interactive = true;
        } else if (arg == "--no-animation") {
            params.use_animation = false;
        } else if (arg == "--no-saves") {
            params.no_saves = true;
        } else if (arg == "-s" || arg == "--seed") {
            params.seed = static_cast<int32_t>(std::stoi(argv[++i]));
        } else if (arg == "-t" || arg == "--threads") {
            params.n_threads = static_cast<int32_t>(std::stoi(argv[++i]));
        } else if (arg == "-p" || arg == "--prompt") {
            params.prompt = argv[++i];
        } else if (arg == "--random-prompt") {
            params.prompt = random_prompt(params.seed);
        } else if (arg == "-n" || arg == "--n_predict") {
            params.n_predict = static_cast<int32_t>(std::stoi(argv[++i]));
        } else if (arg == "--top_k") {
            params.top_k = static_cast<int32_t>(std::stoi(argv[++i]));
        } else if (arg == "--top_p") {
            params.top_p = static_cast<float>(std::stof(argv[++i]));
        } else if (arg == "--temp") {
            params.temp = static_cast<float>(std::stof(argv[++i]));
        } else if (arg == "-b" || arg == "--batch_size") {
            params.n_batch = static_cast<int32_t>(std::stoi(argv[++i]));
        } else if (arg == "--n_ctx") {
            params.n_ctx = static_cast<int>(std::stoi(argv[++i]));
        } else if (arg == "--repeat_penalty") {
            params.repeat_penalty = static_cast<float>(std::stof(argv[++i]));
        } else if (arg == "--repeat_last_n") {
            params.repeat_last_n = static_cast<int>(std::stoi(argv[++i]));
        } else if (arg == "--context_erase") {
            params.context_erase = static_cast<float>(std::stof(argv[++i]));
        } else if (arg == "--b_token") {
            params.b_token = argv[++i];
        } else if (arg == "--e_token") {
            params.e_token = argv[++i];
        } else if (arg == "--load_template") {
            params.load_template = argv[++i];
        } else if (arg == "--save_log") {
            params.save_log = argv[++i];
        } else if (arg == "--load_log") {
            params.load_log = argv[++i];
        } else if (arg == "--save_dir") {
            params.save_dir = argv[++i];    
        } else if (arg == "--save_name") {
            params.save_name = argv[++i];
        } else if (arg == "-m" || arg == "--model") {
            params.model = argv[++i];
        } else if (arg == "-h" || arg == "--help") {
            print_usage(argc, argv, params);
            exit(0);
        } else if (arg == "-v" || arg == "--version") {
            print_version();
            exit(0);
        } else {
            fprintf(stderr, "error: unknown argument: %s\n", arg.c_str());
            print_usage(argc, argv, params);
            exit(0);
        }
    }
    //get path to program
    params.path = pathname_directory(argv[0]);
	params.path.append("/");

	
    return true;
}


#endif
